### PR TITLE
Add Polygon centroid and center of mass

### PR DESCRIPTION
**Pull request by nighthawk**
This ports Turf's [`centerOfMass`](http://turfjs.org/docs/#centerOfMass) and [`centroid`](http://turfjs.org/docs/#centroid) to Swift, at least for the individual `Polygon` case. I've followed the same implementation and test cases as from the JS repo. 

Note: This could be extended to support the other geometries that Turf.js supports, though this should then also add support for convex hull calculations.

This addresses part of #145.

### CODE REVIEW

File: Sources/Turf/CoreLocation.swift [hunk 1]
```diff
@@ -54,12 +54,12 @@ public struct LocationCoordinate2D {
     /**
      The latitude in degrees.
      */
-    public let latitude: LocationDegrees
+    public internal(set) var latitude: LocationDegrees
```
Comment (1ec5) [inline Sources/Turf/CoreLocation.swift:57]: It looks like CLLocationCoordinate2D declares these properties as variable even for their getters, so ideally this Linux-specific declaration would be consistent with it.

File: Sources/Turf/Geometries/Polygon.swift [hunk 2]
```diff
@@ -297,4 +297,55 @@ extension Polygon {
             ring[2].longitude == ring[0].longitude
         )
     }
-}
+
+
+    /// Calculates the centroid using the mean of all vertices.
+    /// This lessens the effect of small islands and artifacts when calculating the centroid of a set of polygons.
+    public func centroid() -> LocationCoordinate2D? {
```
Comment (1ec5) [inline Sources/Turf/Geometries/Polygon.swift:304]: Polygon already has `area` as a computed property rather than a method, so having `centroid` and `centerOfMass` be computed properties too would be more elegant. There’s probably a point at which a parameterless operation is expensive enough that the method syntax is more appropriate, but `centroid` ends up being a single reduction, which is similar to `area`.

File: Sources/Turf/Geometries/Polygon.swift [hunk 3]
```diff
@@ -297,4 +297,55 @@ extension Polygon {
...
+        // We take any point to translate all the points around 0
+        guard let center = centroid() else { return nil }
+        let coordinates = outerRing.coordinates
+        let neutralized = coordinates.map {
+            LocationCoordinate2D(latitude: $0.latitude - center.latitude, longitude: $0.longitude - center.longitude)
+        }
+        
+        var signedArea: Double = 0
+        var sum = LocationCoordinate2D(latitude: 0, longitude: 0)
+        let zipped = zip(neutralized.prefix(upTo: neutralized.count - 1), neutralized.suffix(from: 1))
```
Comment (1ec5) [inline Sources/Turf/Geometries/Polygon.swift:329]: Good use of `zip(_:_:)`!

File: Sources/Turf/Geometries/Polygon.swift [hunk 4]
```diff
@@ -297,4 +297,55 @@ extension Polygon {
...
+
+
+    /// Calculates the centroid using the mean of all vertices.
+    /// This lessens the effect of small islands and artifacts when calculating the centroid of a set of polygons.
+    public func centroid() -> LocationCoordinate2D? {
+        let coordinates = outerRing.coordinates.dropLast()
+        let summed = coordinates
+            .reduce(into: LocationCoordinate2D(latitude: 0, longitude: 0)) { acc, next in
+                acc.latitude += next.latitude
+                acc.longitude += next.longitude
```
Comment (1ec5) [inline Sources/Turf/Geometries/Polygon.swift:309]: Should we be concerned about features that span the antimeridian? Technically, GeoJSON doesn’t allow a single feature to span the antimeridian, but it’s common (at least among Mapbox products) to represent it as a single feature with coordinates that extend beyond ±180°. If this implementation doesn’t handle that case correctly, that’s OK; we can open a separate issue for it.

File: Sources/Turf/Geometries/Polygon.swift [hunk 5]
```diff
@@ -297,4 +297,55 @@ extension Polygon {
             ring[2].longitude == ring[0].longitude
         )
     }
-}
+
+
+    /// Calculates the centroid using the mean of all vertices.
+    /// This lessens the effect of small islands and artifacts when calculating the centroid of a set of polygons.
```
Comment (1ec5) [inline Sources/Turf/Geometries/Polygon.swift:303]: I see you’ve added permalinks to the JavaScript tests that you ported; can you add a similar permalink to the implementation that you ported too? That way we can easily update the implementation if anything changes in the JavaScript version.

File: Sources/Turf/Geometries/Polygon.swift [hunk 6]
```diff
@@ -297,4 +297,55 @@ extension Polygon {
...
+    public func centroid() -> LocationCoordinate2D? {
+        let coordinates = outerRing.coordinates.dropLast()
+        let summed = coordinates
+            .reduce(into: LocationCoordinate2D(latitude: 0, longitude: 0)) { acc, next in
```
Comment (1ec5) [inline Sources/Turf/Geometries/Polygon.swift:307]: The JavaScript implementation [relies on the feature’s bounding box](https://github.com/Turfjs/turf/blob/ae066ae320a5e6a95960f2ed608d79d52b7b62d8/packages/turf-center/index.ts#L40). I’d imagine that this implementation is a bit more efficient than the one in `BoundingBox(from:)`. Perhaps we could replace it outright?

File: Sources/Turf/Geometries/Polygon.swift [hunk 7]
```diff
@@ -297,4 +297,55 @@ extension Polygon {
             ring[2].longitude == ring[0].longitude
         )
     }
-}
+
+
+    /// Calculates the centroid using the mean of all vertices.
+    /// This lessens the effect of small islands and artifacts when calculating the centroid of a set of polygons.
+    public func centroid() -> LocationCoordinate2D? {
+        let coordinates = outerRing.coordinates.dropLast()
```
Comment (1ec5) [inline Sources/Turf/Geometries/Polygon.swift:305]: In the future, once similar calculations are implemented for other geometry types, Geometry itself could have `centroid` and `centerOfMass` properties that pass through to type-specific implementations like this.

**Review by 1ec5 — COMMENTED** [review]
Thank you for porting the JavaScript implementation of these methods – I’m sure a lot of developers will find them quite handy. I haven’t gone line by line comparing the port to the original JavaScript, but let me know if you’d like a second pair of eyes on that.

File: Sources/Turf/Geometries/Polygon.swift (hunk 6 above)
Comment (nighthawk) [reply Sources/Turf/Geometries/Polygon.swift:307]: They are different things. The JS implementation that you linked is for `center`. I haven't ported that one, but rather `centroid` (as `centerOfMass` depends on that). Both have different advantages and disadvantages. The bounding boxes `center` is susceptible to outliers, while `centroid` is susceptive to inbalanced polygons.

See for example:
- center: https://github.com/Turfjs/turf/blob/master/packages/turf-center/test/out/imbalanced-polygon.geojson
- centroid: https://github.com/Turfjs/turf/blob/master/packages/turf-centroid/test/out/imbalanced-polygon.geojson

I can add `center`, too, as it's pretty simple.

File: Sources/Turf/CoreLocation.swift (hunk 1 above)
Comment (nighthawk) [reply Sources/Turf/CoreLocation.swift:57]: Good point. Done.

File: Sources/Turf/Geometries/Polygon.swift (hunk 2 above)
Comment (nighthawk) [reply Sources/Turf/Geometries/Polygon.swift:304]: Makes sense. I've renamed the two for consistency.

File: Sources/Turf/Geometries/Polygon.swift (hunk 4 above)
Comment (nighthawk) [reply Sources/Turf/Geometries/Polygon.swift:309]: Good point. I don't think it handles that properly. It should check if the final longitude extends beyond ±180° and then snap it back within the allowed range. Agree that it might be better to create a separate issue for that. (Turf.js might also need fixing for that.)

File: Sources/Turf/Geometries/Polygon.swift (hunk 5 above)
Comment (nighthawk) [reply Sources/Turf/Geometries/Polygon.swift:303]: Sure thing. Done.

File: Sources/Turf/Geometries/Polygon.swift (hunk 4 above)
Comment (1ec5) [reply Sources/Turf/Geometries/Polygon.swift:309]: The fix would be to wrap the longitude to the ±180° range. Snapping would result in truncating the geometry at the antimeridian, which would be unexpected.

**Comment by 1ec5** [comment]
@nighthawk, feel free to push up a revised version when you’re ready, or let us know if you need help. FYI, we’re planning to release v2.0.0 in the coming weeks, but these changes are backwards-compatible, so we can land them in a subsequent minor version if the timing works out better.

File: Sources/Turf/Geometries/Polygon.swift (hunk 4 above)
Comment (nighthawk) [reply Sources/Turf/Geometries/Polygon.swift:322]: Done!

**Comment by nighthawk** [comment]
@1ec5 , no worries. I've pushed a revised version. Please have a look.

A little curiosity that I encountered when implementing `center`: Doing the manual calculation of the latitude and longitude of the bounding boxes center as commited leads to a different result that doing `mid(bbox.southWest, bbox.northEast)`. For the Lyon example it was off by 5 metres. Is that due to the curvature of the Earth? I went with the manual calculation for consistency with Turf.js.

File: Sources/Turf/Geometries/Polygon.swift [hunk 8]
```diff
@@ -298,27 +298,42 @@ extension Polygon {
...
             .reduce(into: LocationCoordinate2D(latitude: 0, longitude: 0)) { acc, next in
                 acc.latitude += next.latitude
                 acc.longitude += next.longitude
             }
-        return .init(
+        return LocationCoordinate2D(
             latitude: summed.latitude / Double(coordinates.count),
             longitude: summed.longitude / Double(coordinates.count)
-        )
+        ).normalized
```
Comment (1ec5) [inline Sources/Turf/Geometries/Polygon.swift:327]: I was hasty in suggesting “wrapping” above: it’s more correct than snapping but maybe not what a developer would expect. If a feature (such as Alaska) straddles the antimeridian, I think this will cause the bounding box to include everything except Alaska at the same latitudes, effectively going the longer way around the world. It could be that your original code worked fine, but a test would give us more confidence. If not, I think it’s OK to restore the original version (which matches Turf.js more closely) and track this edge case separately, as with #118.

File: README.md [hunk 9]
```diff
@@ -75,6 +75,8 @@ Turf.js | Turf-swift
 [turf-bearing](https://turfjs.org/docs/#bearing) | `CLLocationCoordinate2D.direction(to:)`<br>`LocationCoordinate2D.direction(to:)` on Linux<br>`RadianCoordinate2D.direction(to:)`
 [turf-bezier-spline](https://github.com/Turfjs/turf/tree/master/packages/turf-bezier-spline/) | `LineString.bezier(resolution:sharpness:)`
 [turf-boolean-point-in-polygon](https://github.com/Turfjs/turf/tree/master/packages/turf-boolean-point-in-polygon) | `Polygon.contains(_:ignoreBoundary:)`
+[turf-center-of-mass](http://turfjs.org/docs/#centerOfMass) | `Polygon.centerOfMass()` |
+[turf-centroid](http://turfjs.org/docs/#centroid) | `Polygon.centroid()` |
```
Comment (1ec5) [inline README.md:79]:
```suggestion
[turf-center](http://turfjs.org/docs/#center) | `Polygon.center`
[turf-center-of-mass](http://turfjs.org/docs/#centerOfMass) | `Polygon.centerOfMass`
[turf-centroid](http://turfjs.org/docs/#centroid) | `Polygon.centroid`
```

File: Tests/TurfTests/PolygonTests.swift [hunk 10]
```diff
@@ -142,6 +142,153 @@ class PolygonTests: XCTestCase {
 
         XCTAssertEqual(expectedDiameter, diameter, accuracy: 0.25)
     }
+    
+    func testPolygonCentre() {
+        // Adopted from https://github.com/Turfjs/turf/blob/3b20c568e5638f680cde39c26b56fbcf034133f2/packages/turf-center/test.js
+        let coordinate = LocationCoordinate2D(latitude: 45.7536760235992, longitude: 4.841880798339844)
```
Comment (1ec5) [inline Tests/TurfTests/PolygonTests.swift:148]: I suppose the Lyon example would be more offset than expected when using `mid(_:_:)` because of a longstanding issue in which `LocationCoordinate2D.distance(to:)` uses a different approximation of the radius of the Earth than Turf.js does these days: #26.

**Review by 1ec5 — COMMENTED** [review]
Thank you for porting the full complement of centroid-related methods. Just to keep things simple for now, let’s roll back the changes related to the antimeridian and track that issue separately. Otherwise, this PR will be good to go. Thanks!

File: Sources/Turf/Geometries/Polygon.swift (hunk 8 above)
Comment (1ec5) [reply Sources/Turf/Geometries/Polygon.swift:327]: Oh, my mistake, `normalized` is being applied to the final result, not each individual coordinate. So this will do the right thing. Sorry for the confusion.

**Review by 1ec5 — APPROVED** [review]
Ignore my previous review above, this is actually correct. 😅
